toolkit/tools/scheduler/scheduler.go

```diff
@@ -457,6 +476,53 @@ func updateGraphWithImplicitProvides(res *schedulerutils.BuildResult, pkgGraph *
 	return
 }
 
+// setAssociatedDeltaPaths sets the DeltaPath for all of the request's ancillary nodes (both build and run) to the actual
+// RPM paths. A delta node will normally point at the cached RPM path, but we want to point it at the actual RPM if we built it.
+func setAssociatedDeltaPaths(res *schedulerutils.BuildResult, builtFiles []string, pkgGraph *pkggraph.PkgGraph, graphMutex *sync.RWMutex) (err error) {
+	graphMutex.Lock()
+	defer graphMutex.Unlock()
+
+	if !res.WasDelta {
+		err = fmt.Errorf("setAssociatedDeltaPaths called on non-delta node: %s", res.Node.FriendlyName())
+		return
+	}
+
+	// Build map of basename to full path for built files. This will allow us to find the actual RPM path we built for
+	// any given .rpm file built from our ancillary nodes.
+	builtFileMap := make(map[string]string)
+	for _, builtFile := range builtFiles {
+		// We should not expect to see multiple built files with the same basename
+		_, hasConflict := builtFileMap[filepath.Base(builtFile)]
+		if hasConflict {
+			err = fmt.Errorf("multiple built files with same basename: %s", filepath.Base(builtFile))
+			return
+		}
+		logger.Log.Tracef("Built delta file: %s", builtFile)
+		builtFileMap[filepath.Base(builtFile)] = builtFile
+	}
+
+	// Now we can scan for all the run nodes that use the cached RPM path and update them to the actual RPM path.
+	for _, node := range pkgGraph.AllRunNodes() {
```

Comment:
Going through all run nodes every time we encounter a delta build node feels like an overkill. Do we really care about all run nodes? Could we only go through run nodes, which have an edge pointing to any of the build nodes inside `res.AncillaryNodes`?

Btw. it might be worth adding some execution time monitoring to this function.

Reply:
The function is generally quite fast, and while it is possible to do direct edge lookups in the graph it requires switching to the upstream graph packge's interface. I don't think it is worth the increase in complexity and loss of readability here.